### PR TITLE
feat: add deposit functionality to Gateway contract

### LINKED ISSUE #45: Provider deposit for onramping on Gateway contract

### PR DESCRIPTION
**Pull request by OnahProsperity**
This commit adds a new function `deposit` to the Gateway contract, allowing providers to deposit assets. The function requires the amount to be greater than 0, and the asset to be supported. It also requires the provider to approve the Gateway contract on the token of at least the specified amount before calling the function. The function transfers the specified amount of tokens from the sender to the Gateway contract and updates the deposited balances accordingly. The commit also includes tests for the deposit functionality.

closes #45

### CODE REVIEW

File: contracts/Gateway.sol [hunk 1]
```diff
@@ -243,4 +265,9 @@ contract Gateway is IGateway, GatewaySettingManager, PausableUpgradeable {
 	function getFeeDetails() external view returns (uint64, uint256) {
 		return (protocolFeePercent, MAX_BPS);
 	}
+
+	/** @dev See {getProviderStakedBalance-IGateway}. */
```
Comment (chibie) [inline contracts/Gateway.sol:269]: why this? there's no interface like this

File: contracts/Gateway.sol [hunk 2]
```diff
@@ -243,4 +265,9 @@ contract Gateway is IGateway, GatewaySettingManager, PausableUpgradeable {
...
+	function getProviderDepositBalance(address _token, address _provider) external view returns (uint256) {
```
Comment (chibie) [inline contracts/Gateway.sol:270]: let's call this `getBalance`

File: contracts/interfaces/IGateway.sol [hunk 3]
```diff
@@ -141,6 +149,30 @@ interface IGateway {
 	 */
 	function refund(uint256 _fee, bytes32 _orderId) external returns (bool);
 
+	/**
+	 * @notice Allow Provider to deposit assets.
+	 * @dev Reqirements:
```
Comment (chibie) [inline contracts/interfaces/IGateway.sol:154]: ~Reqirements~ Requirements

File: contracts/interfaces/IGateway.sol [hunk 4]
```diff
@@ -141,6 +149,30 @@ interface IGateway {
...
+	 * @param _orderId The ID of the transaction.
+	 * @param _signature The signature of the provider.
+	 * @param _provider The address of the provider.
+	 * @param _senderAddress The address of the sender.
+	 * @param _token The address of the asset.
+	 * @param _amount The amount to be transferred.
+	 * @return bool the withdrawal is successful.
+	 */
+
+
```
Comment (chibie) [inline contracts/interfaces/IGateway.sol:175]: there's no function here?

File: contracts/interfaces/IGateway.sol [hunk 5]
```diff
@@ -161,4 +193,12 @@ interface IGateway {
 	 * @return max_bps The maximum basis points.
 	 */
 	function getFeeDetails() external view returns (uint64 protocolReward, uint256 max_bps);
+
+	/**
+	 * @notice Gets provider staked balance.
+	 * @param _provider The address of the provider.
+	 * @param _asset The address of the asset.
+	 * @return uint256 The staked balance of the provider.
+	 */
```
Comment (chibie) [inline contracts/interfaces/IGateway.sol:202]: remove "staked"... balance is simple enough

File: contracts/interfaces/IGateway.sol [hunk 6]
```diff
@@ -161,4 +193,12 @@ interface IGateway {
...
+	function getProviderDepositBalance(address _asset, address _provider) external view returns (uint256);
```
Comment (chibie) [inline contracts/interfaces/IGateway.sol:203]: function should be named `getBalance`

**Review by chibie — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IGateway.sol (hunk 4 above)
Comment (OnahProsperity) [reply contracts/interfaces/IGateway.sol:175]: The function is meant to be for escrow function

**Review by chibie — APPROVED** [review]

**Pull request merged**